File: android/src/main/kotlin/noise/protocol/CryptoResolver.kt
package noise.protocol

/**
 * Resolves Noise Protocol algorithm names to concrete implementations.
 *
 * The default resolver supports all standard Noise algorithms. Custom resolvers
 * can be created via [DefaultCryptoResolver.Builder] to register additional or
 * replacement algorithms.
 *
 * @see CryptoSuite
 * @see DefaultCryptoResolver
 */
interface CryptoResolver {
    /**
     * Resolves DH, cipher, and hash names into a [CryptoSuite].
     *
     * @param dhName DH function name (e.g. "25519", "448")
     * @param cipherName Cipher name (e.g. "ChaChaPoly", "AESGCM")
     * @param hashName Hash name (e.g. "SHA256", "BLAKE2b")
     * @throws NoiseException.InvalidPattern if any name is not recognized
     */
    fun resolve(dhName: String, cipherName: String, hashName: String): CryptoSuite

    companion object {
        /** Pre-wired resolver with all standard Noise algorithms. */
        val default: CryptoResolver = DefaultCryptoResolver.Builder()
            .dh("25519") { Curve25519DH }
            .dh("448") { X448DH }
            .cipher("ChaChaPoly") { ChaChaPoly }
            .cipher("AESGCM") { AESGCM }
            .hash("SHA256") { SHA256Hash }
            .hash("SHA512") { SHA512Hash }
            .hash("BLAKE2b") { Blake2bHash }
            .hash("BLAKE2s") { Blake2sHash }
            .build()
    }
}

/**
 * Registry-backed [CryptoResolver] built via [Builder].
 *
 * Immutable after construction — safe to share across threads.
 */
class DefaultCryptoResolver private constructor(
    private val dhRegistry: Map<String, () -> DH>,
    private val cipherRegistry: Map<String, () -> CipherFunction>,
    private val hashRegistry: Map<String, () -> HashFunction>
) : CryptoResolver {

    override fun resolve(dhName: String, cipherName: String, hashName: String): CryptoSuite {
        val dh = dhRegistry[dhName]?.invoke()
            ?: throw NoiseException.InvalidPattern("Unsupported DH: $dhName")
        val cipher = cipherRegistry[cipherName]?.invoke()
            ?: throw NoiseException.InvalidPattern("Unsupported cipher: $cipherName")
        val hash = hashRegistry[hashName]?.invoke()
            ?: throw NoiseException.InvalidPattern("Unsupported hash: $hashName")
        return CryptoSuite(dh, cipher, hash)
    }

    /** Builds a [DefaultCryptoResolver] by registering algorithm factories. */
    class Builder {
        private val dh = mutableMapOf<String, () -> DH>()
        private val cipher = mutableMapOf<String, () -> CipherFunction>()
        private val hash = mutableMapOf<String, () -> HashFunction>()

        fun dh(name: String, factory: () -> DH) = apply { dh[name] = factory }
        fun cipher(name: String, factory: () -> CipherFunction) = apply { cipher[name] = factory }
        fun hash(name: String, factory: () -> HashFunction) = apply { hash[name] = factory }

        fun build(): DefaultCryptoResolver =
            DefaultCryptoResolver(dh.toMap(), cipher.toMap(), hash.toMap())
    }
}

File: ios/Sources/NoiseProtocol/CryptoResolver.swift
import Foundation

/// Resolves Noise Protocol algorithm names to concrete implementations.
public protocol CryptoResolver: Sendable {
    func resolve(dhName: String, cipherName: String, hashName: String) throws -> CryptoSuite
}

/// Registry-backed CryptoResolver built via Builder.
public final class DefaultCryptoResolver: CryptoResolver {
    private let dhRegistry: [String: @Sendable () -> DH]
    private let cipherRegistry: [String: @Sendable () -> CipherFunction]
    private let hashRegistry: [String: @Sendable () -> HashFunction]

    private init(dhRegistry: [String: @Sendable () -> DH],
                 cipherRegistry: [String: @Sendable () -> CipherFunction],
                 hashRegistry: [String: @Sendable () -> HashFunction]) {
        self.dhRegistry = dhRegistry
        self.cipherRegistry = cipherRegistry
        self.hashRegistry = hashRegistry
    }

    public func resolve(dhName: String, cipherName: String, hashName: String) throws -> CryptoSuite {
        guard let dhFactory = dhRegistry[dhName] else {
            throw NoiseError.invalidPattern("Unsupported DH: \(dhName)")
        }
        guard let cipherFactory = cipherRegistry[cipherName] else {
            throw NoiseError.invalidPattern("Unsupported cipher: \(cipherName)")
        }
        guard let hashFactory = hashRegistry[hashName] else {
            throw NoiseError.invalidPattern("Unsupported hash: \(hashName)")
        }
        return CryptoSuite(dh: dhFactory(), cipher: cipherFactory(), hash: hashFactory())
    }

    /// Builds a DefaultCryptoResolver by registering algorithm factories.
    public final class Builder {
        private var dhMap: [String: @Sendable () -> DH] = [:]
        private var cipherMap: [String: @Sendable () -> CipherFunction] = [:]
        private var hashMap: [String: @Sendable () -> HashFunction] = [:]

        public init() {}

        @discardableResult
        public func dh(_ name: String, factory: @escaping @Sendable () -> DH) -> Builder {
            dhMap[name] = factory; return self
        }

        @discardableResult
        public func cipher(_ name: String, factory: @escaping @Sendable () -> CipherFunction) -> Builder {
            cipherMap[name] = factory; return self
        }

        @discardableResult
        public func hash(_ name: String, factory: @escaping @Sendable () -> HashFunction) -> Builder {
            hashMap[name] = factory; return self
        }

        public func build() -> DefaultCryptoResolver {
            DefaultCryptoResolver(dhRegistry: dhMap, cipherRegistry: cipherMap, hashRegistry: hashMap)
        }
    }

    /// Pre-wired resolver with all standard Noise algorithms.
    public static let `default`: CryptoResolver = DefaultCryptoResolver.Builder()
        .dh("25519") { Curve25519DH() }
        .dh("448") { X448DH_() }
        .cipher("ChaChaPoly") { ChaChaPoly_() }
        .cipher("AESGCM") { AESGCM_() }
        .hash("SHA256") { SHA256Hash_() }
        .hash("SHA512") { SHA512Hash_() }
        .hash("BLAKE2b") { Blake2bHash_() }
        .hash("BLAKE2s") { Blake2sHash_() }
        .build()
}
